cli2telegram/daemon.py

```diff
@@ -0,0 +1,132 @@
+        signal(SIGTERM, self.signal_handler) # systemctl stop
+
+        threading.Thread(target=self.send_message_worker, daemon=True).start()
+
+        LOGGER.info("daemon listening on {}".format(self.PIPE_NAME))
+
+        while True:
+            with open(self.PIPE_NAME) as pipe:
+                LOGGER.debug("wating for input from pipe")
+                pipe_buffer = ''
+                while True:
+                    data = pipe.read()
+                    LOGGER.debug(f"received {data}")
+                    if len(data) == 0 and self.RUNNING:
+                        LOGGER.debug("No data from pipe. Sending buffer to telegram.");
+                        self.message_queue.put(pipe_buffer)
+                        pipe_buffer = ''
+                        break
+                    elif not self.RUNNING:
+                        break
+                    else:
+                        pipe_buffer += data
+            if not self.RUNNING:
+                break
+
+    def send_message_worker(self):
+        while True:
+            message = self.message_queue.get()
+            LOGGER.debug(f"Got {message} from queue. sending...")
+            self._try_send_message(message)
+            self.message_queue.task_done()
+
+    def _try_send_message(self, message: str):
+        """
+        Sends a message
+        :param message: the message to send
+        """
+        started_trying = datetime.now()
+        success = False
+        updater = Updater(token=self.CONFIG.TELEGRAM_BOT_TOKEN.value, use_context=True)
+        while not success:
+            try:
+                chat_id = self.CONFIG.TELEGRAM_CHAT_ID.value
+                send_message(updater.bot, chat_id, message, parse_mode="markdown")
+                success = True
+            except Exception as ex:
+                LOGGER.exception(ex)
+
+                if not self.CONFIG.RETRY_ENABLED.value:
+                    break
+
+                tried_for = datetime.now() - started_trying
+                if tried_for > self.CONFIG.RETRY_GIVE_UP_AFTER.value:
+                    LOGGER.warning(f"Giving up after trying for: {tried_for}. Placing message back in queue.")
+                    self.message_queue.put(message)
+                    break
+
+                timeout_seconds = self.CONFIG.RETRY_TIMEOUT.value.total_seconds()
+                LOGGER.error(f"Error sending message, retrying in {timeout_seconds} seconds...")
+                time.sleep(timeout_seconds)
+
+    def signal_handler(self, signal_received, frame):
+        self.RUNNING = False
+        
+        LOGGER.debug(f"Caught {signal_received}. Attempting to send all messages from queue.")
+        
+        try:
+            os.unlink(self.PIPE_NAME)
+        except OSError as ose:
+            if os.path.exists(self.PIPE_NAME):
+                LOGGER.warning("Unable to clean up named pipe")
+                raise
+        
+        self.message_queue.join()
```

Comment:
Hmm not sure about this... The idea it good, try to send the stuff we have before exiting, but depending on the `RETRY_GIVE_UP_AFTER` config and the amount of items in the queue, this can potentially take a very long time. Maybe there is an easy way to limit the `join()` to something like 5 seconds and kill the thread if it takes longer?

Reply:
Yes. Since we're aren't adding failed messages back to the queue, your first point won't have a compounding impact. 

Regarding,
`self.message_queue.join()`
limiting the join will look like this:
```
stop = time() + 5
while self.message_queue.unfinished_tasks and time() < stop:
    sleep(1)
```
Also, since the message worker thread is created with `daemon=True`, it will terminate when the main thread terminates.